core/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java

```diff
@@ -185,7 +185,13 @@ private void performOnReplica(final ShardRouting shard, final ReplicaRequest rep
             @Override
             public void onResponse(ReplicaResponse response) {
                 successfulShards.incrementAndGet();
-                primary.updateLocalCheckpointForShard(response.allocationId(), response.localCheckpoint());
+                try {
+                    primary.updateLocalCheckpointForShard(response.allocationId(), response.localCheckpoint());
+                } catch (final Exception e) {
```

Comment:
I think we need to be careful hare. It's ok to have closed the shard. For example if the index was deleted (or more complicated, if this a target primary and the source node left the cluster before the master activated this primary. In these cases, we shouldn't fail the operation (nor the shard, for that matter) but rather let things finish in their natural flow. If the op was properly replicated to all active shards, we're all good. If not, we're in trouble but that would be picked up by the logic handling failures from replicas.